### PR TITLE
Restore environment after `make` to avoid mixing-up linker flags

### LINKED ISSUE #29983: `vcpkg_build_make` fails to restore empty `_LINK_` environment variable
`vcpkg_build_make` saves the value of the `_LINK_` environment variable to `config_link_backup` before modifying it, with the intention of restoring the original value after running `make` :
https://github.com/microsoft/vcpkg/blob/8e82d67b78a6c7fe93bf25a3bfb595f01b2d7d7e/scripts/cmake/vcpkg_build_make.cmake#L130
Unfortunately, this will only happen if `_LINK_` is non-empty to begin with, because the branch that restores the original value is guarded by `if(config_link_backup)`:
https://github.com/microsoft/vcpkg/blob/8e82d67b78a6c7fe93bf25a3bfb595f01b2d7d7e/scripts/cmake/vcpkg_build_make.cmake#L174-L175
As a result, `release` builds of `make`-backed ports will see the accumulated linker flags of the `debug` and `release` configurations, since `release` is built after `debug`, and the empty value of `_LINK_` is not restored after the build of `debug`.

**Environment**
- OS: Windows
- Compiler: MSVC

https://github.com/microsoft/vcpkg/blob/8e82d67b78a6c7fe93bf25a3bfb595f01b2d7d7e/scripts/cmake/vcpkg_build_make.cmake#L106-L107

**To Reproduce**
1. `./vcpkg.exe install icu:x64-windows` (fixed triplet chosen for simplicity, should work with other triplets as well)
2. Review the output of `grep -m 1 -A 4 '^LINK\.EXE' buildtrees/icu/build-x64-windows-*-out.log`
3. Notice how the `debug` flags "spill" into the `release` flags

```
buildtrees/icu/build-x64-windows-dbg-out.log:LINK.EXE -DLL -Xlinker -Xlinker -Xlinker -machine:x64 -Xlinker -Xlinker -Xlinker -nologo -Xlinker -Xlinker -Xlinker -debug -Xlinker -Xlinker -Xlinker -INCREMENTAL  -nologo -NOENTRY -IMPLIB:icudtd.lib -out:icudtd72.dll stubdata.o -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -lcomdlg32 -ladvapi32
buildtrees/icu/build-x64-windows-dbg-out.log-Microsoft (R) Incremental Linker Version 14.35.32215.0
buildtrees/icu/build-x64-windows-dbg-out.log-Copyright (C) Microsoft Corporation.  All rights reserved.
buildtrees/icu/build-x64-windows-dbg-out.log-
buildtrees/icu/build-x64-windows-dbg-out.log--machine:x64 -nologo -debug -INCREMENTAL
--
buildtrees/icu/build-x64-windows-rel-out.log:LINK.EXE -DLL -Xlinker -Xlinker -Xlinker -machine:x64 -Xlinker -Xlinker -Xlinker -nologo -Xlinker -Xlinker -Xlinker -DEBUG -Xlinker -Xlinker -Xlinker -INCREMENTAL:NO -Xlinker -Xlinker -Xlinker -OPT:REF -Xlinker -Xlinker -Xlinker -OPT:ICF  -nologo -NOENTRY -IMPLIB:icudt.lib -out:icudt72.dll stubdata.o -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -lcomdlg32 -ladvapi32
buildtrees/icu/build-x64-windows-rel-out.log-Microsoft (R) Incremental Linker Version 14.35.32215.0
buildtrees/icu/build-x64-windows-rel-out.log-Copyright (C) Microsoft Corporation.  All rights reserved.
buildtrees/icu/build-x64-windows-rel-out.log-
buildtrees/icu/build-x64-windows-rel-out.log--machine:x64 -nologo -debug -INCREMENTAL -machine:x64 -nologo -DEBUG -INCREMENTAL:NO -OPT:REF -OPT:ICF
```

**Expected behavior**
```
buildtrees/icu/build-x64-windows-dbg-out.log:LINK.EXE -DLL -Xlinker -Xlinker -Xlinker -machine:x64 -Xlinker -Xlinker -Xlinker -nologo -Xlinker -Xlinker -Xlinker -debug -Xlinker -Xlinker -Xlinker -INCREMENTAL  -nologo -NOENTRY -IMPLIB:icudtd.lib -out:icudtd72.dll stubdata.o -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -lcomdlg32 -ladvapi32
buildtrees/icu/build-x64-windows-dbg-out.log-Microsoft (R) Incremental Linker Version 14.35.32215.0
buildtrees/icu/build-x64-windows-dbg-out.log-Copyright (C) Microsoft Corporation.  All rights reserved.
buildtrees/icu/build-x64-windows-dbg-out.log-
buildtrees/icu/build-x64-windows-dbg-out.log--machine:x64 -nologo -debug -INCREMENTAL
--
buildtrees/icu/build-x64-windows-rel-out.log:LINK.EXE -DLL -Xlinker -Xlinker -Xlinker -machine:x64 -Xlinker -Xlinker -Xlinker -nologo -Xlinker -Xlinker -Xlinker -DEBUG -Xlinker -Xlinker -Xlinker -INCREMENTAL:NO -Xlinker -Xlinker -Xlinker -OPT:REF -Xlinker -Xlinker -Xlinker -OPT:ICF  -nologo -NOENTRY -IMPLIB:icudt.lib -out:icudt72.dll stubdata.o -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -lcomdlg32 -ladvapi32
buildtrees/icu/build-x64-windows-rel-out.log-Microsoft (R) Incremental Linker Version 14.35.32215.0
buildtrees/icu/build-x64-windows-rel-out.log-Copyright (C) Microsoft Corporation.  All rights reserved.
buildtrees/icu/build-x64-windows-rel-out.log-
buildtrees/icu/build-x64-windows-rel-out.log--machine:x64 -nologo -DEBUG -INCREMENTAL:NO -OPT:REF -OPT:ICF
```

### PR DESCRIPTION
**Pull request by jre-lsft**
At first I considered just replacing the whole `config_link_backup` logic in `vcpkg_build_make` with `vcpkg_{backup,restore}_env_variables(VARS _LINK_)`.
Unfortunately, I soon realized that the same logic exists in `vcpkg_configure_make`, only that in this case `vcpkg_backup_env_variables(VARS _LINK_)` is already used in the same scope.
Since `vcpkg_{backup,restore}_env_variables()` relies on the enclosing variable scope, which makes nested calls (within the same function) difficult, I kept the `config_link_backup` logic (for now).

Note that thanks to CMake 3.25 and `block()` scopes, nested calls to `vcpkg_{backup,restore}_env_variables()` are now actually an option.
However, even though vcpkg already uses CMake 3.25, I wasn't sure if this meant that backwards compatibility was not an issue.

Fixes #29983.

### CODE REVIEW

**Comment by jre-lsft** [comment]
@microsoft-github-policy-service agree company="elusoft GmbH"

**Comment by Cheney-W** [comment]
You could get the error logs from https://dev.azure.com/vcpkg/public/_build/results?buildId=86250&view=artifacts&pathAsName=false&type=publishedArtifacts

**Comment by jre-lsft** [comment]
Yeah, I really shouldn't have made that second set of changes.
They are not relevant to the issue I was facing and are only related in that the stray `vcpkg_backup_env_vars(VARS _CL_ _LINK_)` caught my attention, as it was the reason I did not to use `vcpkg_{backup,restore}_env_vars()` to save/restore the build configuration specific changes to `_LINK_` in the `foreach()` loops.
Seeing that there was no corresponding `vcpkg_restore_env_vars(VARS _CL_ _LINK_)` I jumped to conclusions.
Now, after re-reading the comments, I realize that the UWP-related modifications to `_CL_` and `_LINK_` are intentionally "leaked".
However, even in that case, it occurs to me that the stray `vcpkg_backup_env_vars(VARS _CL_ _LINK_)` doesn't make much sense, but perhaps someone more familiar with the code base could take a second look.
Anyway, I've removed the irrelevant changes, please forgive my ignorance.

**Review by Cheney-W — APPROVED** [review]

File: scripts/cmake/vcpkg_build_make.cmake [hunk 1]
```diff
@@ -129,6 +129,8 @@ function(vcpkg_build_make)
             if(LINK_ENV_${cmake_buildtype})
                 set(config_link_backup "$ENV{_LINK_}")
                 set(ENV{_LINK_} "${LINK_ENV_${cmake_buildtype}}")
+            else()
+                unset(config_link_backup)
```
Comment (dan-shaw) [inline scripts/cmake/vcpkg_build_make.cmake:133]: Why do we unset this variable if it was not set?

Comment (jre-lsft) [reply scripts/cmake/vcpkg_build_make.cmake:133]: (In the following, I'll assume that both _this variable_ and _it_ are referring to `config_link_backup`. However, if _it_ refers to `LINK_ENV_${cmake_buildtype}`, you'll have to make some mental adjustments.)

But we don't know that, do we?
Note that I have moved the `unset()` up from the second `if()` clause.
As a result, we might observe the value of `config_link_backup` from the first iteration in the second iteration of the `foreach()` loop.
Regardless, and to be pedantic, we don't even know that `config_link_backup` is unset when the `foreach()` loop is first entered, it could exist in the global scope, too.
Call me paranoid.

The goal of this PR is to make sure that the value of the `_LINK_` environment variable is restored in the second `if()` clause **if and only if** it was changed in the first `if()` clause.
There are certainly several ways to implement this logic in a semantically equivalent way.
For example, we could not look at `config_link_backup` at all in the second `if()` condition and instead use the exact same condition as in the first `if()` clause, i.e. `if(LINK_ENV_${cmake_buildtype})` (assuming `LINK_ENV_${cmake_buildtype}` itself is not modified between the first and the second `if()` clause).
I chose not to do this because I felt that it would increase, however slightly, the chances that the two conditions would become out of sync over time, breaking the _if and only if_ requirement.
In other words: I decided to go with the proposed implementation because it (kind of) keeps both decisions, whether to modify `_LINK_` in the first place and whether to restore its value, close together in the first `if()` clause, by encoding the latter decision into `config_link_backup` (it's either `DEFINED` or not).
